[WIP] Add encoder/decoder for Avro object container files

veedo opened the pull request:
For some of my projects, I need full control of the encoding/decoding process and AvroEx provides a good basis for that.
The only things missing from AvroEx that I use a lot are object containers and cloud wire formats.
This pull request is my attempt to add object containers in a flexible way that gives a user maximum control over the process.

Theory of operation:
* Each part of the container can be encoded/decoded separately
* Codec implementation can be supplied by the user
  * Currently using the :avro_ex app config so that they can configure it in their mix config
  * The snappy implementation is provided Without the underlying snappyer because snappy compression does something weird in Avro, but adding snappyer as a dependency adds a NIF compile requirement which is undesirable for cross compilation
  * I believe compression codecs are only used in Object Containers, so I placed them under that module. Let me know if that is not the case.
* Just return the encoded data, let the user decide how they want to write the file
  * Not yet sure how well decoding will work for this concept, might be forced to use IO objects
  * Could be solved by providing functions for figuring out how much data to read for each chunk?


Please provide feedback on the PR as I go in case there's something untenable

veedo committed:
Add encoder for object container files
Only adding the raising functions for now
The implementation is designed so that people can pass different implementations of the codecs as required.
Some people may want to use pure beam functions, some may want to use NIFs.

davydog187 commented:
Thank you for the PR! I will provide some direct feedback this week. One immediate piece of feedback is to move all codecs to separate files

veedo committed:
Add file header encoding tests
Improve how block headers are encoded to use an avro record
This will help during decoding, since longs are variable length

veedo committed:
Split each codec into its own file

davydog187 commented on lib/avro_ex/object_container.ex:
```diff
@@ -13,16 +13,25 @@ defmodule AvroEx.ObjectContainer do
-               ]
-             }
-             """)
+  @bh_schema AvroEx.decode_schema!(~S({
+    "type":"record","name":"block_header",
+    "fields":[
+      {"name":"num_objects","type":"long"},
+      {"name":"num_bytes","type":"long"}
+    ]
+  }))
+  @fh_schema AvroEx.decode_schema!(~S({
+    "type": "record", "name": "org.apache.avro.file.Header",
+    "fields" : [
+      {"name": "magic", "type": {"type": "fixed", "name": "Magic", "size": 4}},
+      {"name": "meta", "type": {"type": "map", "values": "bytes"}},
+      {"name": "sync", "type": {"type": "fixed", "name": "Sync", "size": 16}}
+    ]
+  }))
```
1. [decode_schema!](https://hexdocs.pm/avro_ex/AvroEx.html#decode_schema!/2) supports directly passing elixir terms, so this can pass an elixir map directly rather than a string
2. Calling `decode_schema!/1` here causing a compile time dependency between this module and `AvroEx`. Since this is a premature optimization, I suggest we move out of module attributes

WDYT?

veedo replied:
> 1. [decode_schema!](https://hexdocs.pm/avro_ex/AvroEx.html#decode_schema!/2) supports directly passing elixir terms, so this can pass an elixir map directly rather than a string

Didn't notice that, thanks.
A good reason for using json may be to "match" with the Avro spec:
![image](https://user-images.githubusercontent.com/5835448/185962089-6313c61c-f03b-48ea-aeaf-f9d21eab6d8e.png)
vs
```elixir
 %{
   "type" => "record",
   "name" => "org.apache.avro.file.Header",
   "fields" => [
     %{"name" => "magic", "type" => %{"type" => "fixed", "name" => "Magic", "size" => 4}},
     %{"name" => "meta", "type" => %{"type" => "map", "values" => "bytes"}},
     %{"name" => "sync", "type" => %{"type" => "fixed", "name" => "Sync", "size" => 16}}
   ]
 })
```
the spec is unlikely to change, so no one will ever need to update it
I don't have a strong opinion about it :shrug: 
let me know which you prefer with the options side by side

>     2. Calling `decode_schema!/1` here causing a compile time dependency between this module and `AvroEx`. Since this is a premature optimization, I suggest we move out of module attributes
> 
> 
> WDYT?

I'm not aware of any downsides. I work with a lot of embedded devices though, so doing stuff at compile time feels natural to me.
Is there a reason to avoid this optimization?
I can stick it in the `file_header_schema/1` instead

davydog187 replied:
Since it shouldn't change much over time, I think its fine to represent in elixir and get the benefits of formatting. It is also easier to edit as elixir code

Regarding compile-time, it shouldn't make a massive difference here, but accrued work at compile time does effect the time it takes to compile the library.

davydog187 commented on test/object_container_encode_test.exs:
```diff
@@ -0,0 +1,92 @@
+defmodule AvroEx.ObjectContainer.Encode.Test do
+  use ExUnit.Case, async: true
+
+  @test_module AvroEx.ObjectContainer
```
can you please call the module directly rather than `@test_module`? We have begun to get away from this practice elsewhere, but we haven't fully cleaned up in the codebase
```suggestion
  alias AvroEx.ObjectContainer
```

veedo replied:
Will do, that's the way I prefer it too

davydog187 commented on lib/avro_ex/object_container.ex:
```diff
@@ -0,0 +1,85 @@
+    }
+  end
+
+  def encode_file_header!(ocf = %__MODULE__{}) do
```
```suggestion
  def encode_file_header!(%__MODULE__{} = ocf) do
```

davydog187 commented:
Took a quick look at this, here is some general feedback:

1. To keep the API surface area of `AvroEx` as small as possible, I would suggest that we have top-level APIs for working with Object Container Files in `AvroEx`. We can have the implementation delegate out to the `AvroEx.ObjectContainer` module
2. There was mention of using application configuration for the codec. I would advise against this and instead just allow the user to pass a keyword argument to the library, if the user of the library wants to use Application config let them do that in their own application. See the Elixir [library guidelines](https://hexdocs.pm/elixir/main/library-guidelines.html#avoid-application-configuration)

veedo committed:
Use alias instead of property

veedo commented on lib/avro_ex/object_container.ex:
```diff
@@ -13,16 +13,25 @@ defmodule AvroEx.ObjectContainer do
+    ]
+  }))
+
+  def magic(), do: @magic
+  def block_header_schema(), do: @bh_schema
+  def file_header_schema(), do: @fh_schema
```
The magic is unnecessary, but the raw schemas are useful to have.
I'm using them in tests right now. It feels icky to repeat these schemas in the encoding and decoding test files.
Is there a better way other than exposing them?

davydog187 replied:
IMO we should drop the need for them at tests, since we can just validate against expected inputs/outputs

veedo replied:
That is true. The format isn't expected to change either, so the tests won't need to be updated.
I'll try it out and see how it looks, it may make the tests 99% encoded data blobs though :stuck_out_tongue:

davydog187 replied:
The best way to test this is that you can round trip the data through the encoder, so that should be completely reasonable

veedo committed:
Swap __MODULE__ position in argument pattern
Matches the rest of the library

davydog187 commented on lib/avro_ex/object_container/codec/snappy.ex:
```diff
@@ -0,0 +1,21 @@
+  @behaviour AvroEx.ObjectContainer.Codec
+  @impl AvroEx.ObjectContainer.Codec
+  def encode!(data) do
+    {:ok, compressed} = :snappyer.compress(data)
```
We will need to add these as optional dependencies, no? We should also probably not compile this module if the user does not have `snappyer` installed

veedo replied:
I'm torn over how to handle it. 
The issue is that the snappy codec is implemented differently from all the other codecs.
I haven't looked at how to do it yet, but it would be nice to only compile it in if :snappyer is in the list of compiled/included apps or maybe some kind of build flag?
The implementation could just live in the docs, but it seems to be a popular codec 
What do you think?

My plan was to worry about it after finishing the rest of the object container

veedo commented:
> Took a quick look at this, here is some general feedback:
> 
>     1. To keep the API surface area of `AvroEx` as small as possible, I would suggest that we have top-level APIs for working with Object Container Files in `AvroEx`. We can have the implementation delegate out to the `AvroEx.ObjectContainer` module

will do :+1: 

>     2. There was mention of using application configuration for the codec. I would advise against this and instead just allow the user to pass a keyword argument to the library, if the user of the library wants to use Application config let them do that in their own application. See the Elixir [library guidelines](https://hexdocs.pm/elixir/main/library-guidelines.html#avoid-application-configuration)

Thanks, that is a useful document.
Right now the codec is passed in anyways, i'll just have to think about how the name+implementation will work.
I'll probably just add a `name/0` function to the behaviour

veedo committed:
Use elixir maps instead JSON strings for static schemas

veedo committed:
Pass in codec into encoder instead of using env

veedo committed:
Add initial file header decode function

veedo committed:
Improve error handling and responses
Handle case with missing codec
Add tests for parts of the file header decoding required by the spec

davydog187 commented:
hello @veedo, just wanted to check in on this PR, are you waiting on any review from me, or still a WIP?

veedo commented:
> 

I've just been swamped this month unfortunately.
I'll probably make some progress next week/weekend though 😅

davydog187 commented:
No rush! Just wanted to make sure you weren't waiting on me

veedo committed:
Decode file container objects
Implement all the functions that go into decoding parts of the object container

davydog187 commented:
Hello @veedo! Checking back in here, is there anything I can do to help with this PR? If you're not going to come back to it, we can consider other options

veedo commented:
> 

The swamping has continued unfortunately, and will probably continue for the next 2 months at least.

Currently the encoding part works correctly and consistently. 
My plan was to finish all the tests and start on decoding.
I can split the encoding part out into its own PR and remove all the decoding parts, 
but that may be a bit weird for a user expecting both.

I'll whip myself to finish the encoding tests this weekend.
How would you like to handle it? Splitting shouldn't be too much more work.

veedo committed:
File decoding
Implement the function that decodes an object container

veedo committed:
Test Encode/Decode blocks and object containers

veedo committed:
Merge branch 'master' into object_container_file

veedo committed:
Make the snappy codec optional.
The implementation of the snappy codec in avro is quite
unique, so providing an implementation is valuable.
Uses the optional dependency method similar to ecto.

veedo commented:
@davydog187 Disregard my last comment. I had forgotten how much was left to do.
The encoding and decoding works, I just need to do the documentation.
I'll see if I can get it into an understandable state today.

davydog187 commented:
@veedo wanted to remind you about this PR! Its come so far I'd love to still be able to get this merged at some point

cgarvis commented:
Anyway I can help with this?  Would definitely use it for an Iceberg package I'm building.

veedo committed:
Documentation and specs for the object container codec behaviour

veedo committed:
Document the Avro Ovject Container module.

Fix missing decompress call.

veedo commented:
> Anyway I can help with this? Would definitely use it for an Iceberg package I'm building.

@cgarvis - If you can test some files encoded by an independent source, that would be helpful to have an independent check.

@davydog187 - Added documentation. Everything should be good to go. Let me know if anything is wrong.
I'll be on a work trip for a couple of weeks, so I'll have a lot of free time to fix anything up in the evenings. :laughing:
